compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEUnrollAndDistributeWorkgroup.cpp

```diff
@@ -0,0 +1,511 @@
+    for (Operation *userOp : dmaOp->getUsers()) {
+      if (auto coreOp = userOp->getParentOfType<AMDAIE::CoreOp>()) {
+        auto workgroupOp = coreOp->getParentOfType<AMDAIE::WorkgroupOp>();
+        if (!workgroupOp) continue;
+
+        Attribute sourceMemspace = dmaOp.getSourceObjectFifo().getMemorySpace();
+        Attribute targetMemspace = dmaOp.getTargetObjectFifo().getMemorySpace();
+        if (sourceMemspace &&
+            dyn_cast<IntegerAttr>(sourceMemspace).getInt() == 2) {
+          // Source on L1
+          insertTile(dmaOp.getSourceObjectFifo(),
+                     coreOp.getTileOp().getResult());
+        } else if (targetMemspace &&
+                   dyn_cast<IntegerAttr>(targetMemspace).getInt() == 2) {
+          // Target on L1
+          insertTile(dmaOp.getTargetObjectFifo(),
+                     coreOp.getTileOp().getResult());
+        }
+
+        // Move tile to beginning of workgroup to make sure ssa values are not
+        // dominated
+        Block *workgroupBlock = workgroupOp.getBody();
+        rewriter.moveOpBefore(coreOp.getTileOp(), workgroupBlock,
+                              workgroupBlock->begin());
+      }
+    }
+    return WalkResult::advance();
+  });
+
+  // Update the logical objectfifos with assigned tiles
+  for (auto &&[logicalObjectFifo, tiles] : logicalObjectFifosToTiles) {
+    rewriter.setInsertionPoint(logicalObjectFifo);
+    rewriter.replaceOpWithNewOp<AMDAIE::LogicalObjectFifoFromMemrefOp>(
+        logicalObjectFifo,
+        logicalObjectFifo.getOutput().getType().cast<LogicalObjectFifoType>(),
+        logicalObjectFifo.getMemref(), tiles.takeVector());
+  }
+  return success();
+}
+
+/// Return the tiles of the sources of the users of this logical objectfifo.
+LogicalResult getUserSourceTiles(
+    AMDAIE::LogicalObjectFifoFromMemrefOp logicalObjectFifo,
+    SmallVectorImpl<AMDAIE::TileOp> &tiles) {
+  llvm::SmallSetVector<AMDAIE::TileOp, 16> tileSet;
+  for (Operation *user : logicalObjectFifo->getUsers()) {
+    if (auto dmaOp = dyn_cast<AMDAIE::DmaCpyNdOp>(user);
+        dmaOp && dmaOp.getTargetObjectFifo() == logicalObjectFifo) {
+      ValueRange tileIndices = dmaOp.getSourceObjectFifo().getTiles();
+      // Only fill in tiles when all sources have tiles.
+      if (tileIndices.empty()) return failure();
+      for (Value index : tileIndices)
+        tileSet.insert(dyn_cast<AMDAIE::TileOp>(index.getDefiningOp()));
+    }
+  }
+  tiles = tileSet.takeVector();
+  return success();
+}
+
+/// Return the tiles of the targets of the users of this logical
+/// objectfifo.
+LogicalResult getUserTargetTiles(
+    AMDAIE::LogicalObjectFifoFromMemrefOp logicalObjectFifo,
+    SmallVectorImpl<AMDAIE::TileOp> &tiles) {
+  llvm::SmallSetVector<AMDAIE::TileOp, 16> tileSet;
+  for (Operation *user : logicalObjectFifo->getUsers()) {
+    if (auto dmaOp = dyn_cast<AMDAIE::DmaCpyNdOp>(user);
+        dmaOp && dmaOp.getSourceObjectFifo() == logicalObjectFifo) {
```

Comment:
I guess both of these functions can be combined into one since the only differing thing between the two is `(target, source)` and `(source, target)`.

So maybe you can pass in another variable to the unified function `fetchTarget` ?

And this way the `if` can become :-
```
if (auto dmaOp = dyn_cast<AMDAIE::DmaCpyNdOp>(user)) {
   // Based on `fetchTarget` perform the check and extraction.
   ...
   // Rest of the code remains the same.
}
```